### PR TITLE
chore(scripts): add script to promote mainline to stable

### LINKED ISSUE #12459: Stable Releases: Create new release script for promoting mainline to stable
Part of [Stable Releases](https://www.notion.so/coderhq/Stable-Releases-fcc801d8b9384d94975b2a2b63b20550).

## Overview 

See https://github.com/coder/coder/issues/12458 for context.

In addition to bisecting the current release script to support `stable` and `mainline`, we need to add a new release script to migrate the `stable` label to the latest mainline release. This happens after 2 weeks of fielding the mainline release to the community as our "latest" edge release.

## Requirements

Example usage, promoting 2.9.1 to stable from 2.8.3:
 `promote_stable.sh v2.9.1`

- Accepts the desired version tag to be upgraded
- Migrates `latest` github tag to target version
- Appends "(Stable)" to the target version's release name
- Removes "(Stable)" from previous versions's release name
- `CODER_MAINLINE_VERSION` in install.sh does not need updating

When this action occurs, mainline and stable will point to the same version for 2 weeks. This is acceptable.

### PR DESCRIPTION
**Pull request by mafredri**
Fixes #12459

Example dry-run:

Screenshot 2024-04-23 at 21 16 55

Example dry-run for non-latest version:

Screenshot 2024-04-23 at 21 17 52

**Note:** This PR does not yet update docs to reflect the promoted version. This will be part of #12465.

### CODE REVIEW

**Comment by matifali** [comment]
Do we plan to automate [this](https://coder.com/docs/v2/latest/install/releases#release-schedule) table, too?

**Review by mtojek — APPROVED** [review]
Nice and clean script 👍

File: scripts/release/main.go [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+	}
+
+	// Ensure the version is a valid release.
+	perPage := 20
+	latestReleases, _, err := client.Repositories.ListReleases(ctx, owner, repo, &github.ListOptions{
+		Page:    0,
+		PerPage: perPage,
+	})
+	if err != nil {
+		return err
```
Comment (dannykopping) [inline scripts/release/main.go:116]: I'd suggest wrapping this error or adding a log line before it showing what the current action is

File: scripts/release/main.go [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+	updatedBody = addStableSince(time.Now().UTC(), updatedBody)
+	updatedNewStable.Body = github.String(updatedBody)
+	updatedNewStable.Prerelease = github.Bool(false)
+	updatedNewStable.Draft = github.Bool(false)
+	if !dryRun {
+		_, _, err = client.Repositories.EditRelease(ctx, owner, repo, newStable.GetID(), newStable)
+		if err != nil {
+			return err
+		}
+		logger.Info(ctx, "selected version promoted to stable")
```
Comment (dannykopping) [inline scripts/release/main.go:169]: Could you provide a link to the release in the output for convenience?

File: scripts/release/main_test.go [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+
+Refer to our docs to [install](https://coder.com/docs/v2/latest/install) or [upgrade](https://coder.com/docs/v2/latest/admin/upgrade) Coder, or use a release asset below.
+`,
+		},
+		{
+			name: "WithMainlineBlurb",
+			body: `## Changelog
+
+> [!NOTE]
+> This is a mainline Coder release. We advise enterprise customers without a staging environment to install our [latest stable release](https://github.com/coder/coder/releases/latest) while we refine this version. Learn more about our [Release Schedule](https://coder.com/docs/v2/latest/install/releases).
```
Comment (dannykopping) [inline scripts/release/main_test.go:58]: I think it'd be illustrative / more complete to add a test which shows that multiple lines prefixed with `>` are removed; we'll likely amend this note in the future.

**Review by dannykopping — APPROVED** [review]
Minor notes, LGTM

**Comment by matifali** [comment]
Should these PRs be chore as it's not a user facing feature? It would end up as features in our changelog.

**Comment by mafredri** [comment]
> Should these PRs be chore as it's not a user facing feature? It would end up as features in our changelog.

That’s fair. I’ll change them.

**Comment by mafredri** [comment]
> Do we plan to automate [this](https://coder.com/docs/v2/latest/install/releases#release-schedule) table, too?

Not that I’m aware, it’s not mentioned in #12465.

**Review by dannykopping — APPROVED** [review]
👌

**Comment by mafredri** [comment]
* **#13054** <a href="https://app.graphite.dev/github/pr/coder/coder/13054?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @mafredri and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: scripts/release/main.go [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+	rr := *r
+	return &rr
+}
+
+// addStableSince adds a stable since note to the release body.
+//
+// Example:
+//
+//	> ## Stable (since April 23, 2024)
+func addStableSince(date time.Time, body string) string {
```
Comment (stirby) [inline scripts/release/main.go:187]: Nice!!

**Review by stirby — APPROVED** [review]
Awesome.

**Pull request merged**